modules/adgenerationBidAdapter.js

```diff
@@ -0,0 +1,194 @@
+        bidResponse.native = createNativeAd(body);
+        bidResponse.mediaType = NATIVE;
+      } else {
+        // banner
+        bidResponse.ad = createAd(body, bidRequest);
+      }
+      bidResponses.push(bidResponse);
+    }
+    return bidResponses; // noAdは空を返す
+  },
+
+  /**
+   * Register the user sync pixels which should be dropped after the auction.
+   *
+   * @param {SyncOptions} syncOptions Which user syncs are allowed?
+   * @param {ServerResponse[]} serverResponses List of server's responses.
+   * @return {UserSync[]} The user syncs which should be dropped.
+   */
+  getUserSyncs: function (syncOptions, serverResponses) {
+    const syncs = [];
+    return syncs;
+  }
+};
+
+function createAd(body, bidRequest) {
+  let ad = body.ad;
+  if (body.vastxml && body.vastxml.length > 0) {
+    ad = `<div id="apvad-${bidRequest.bidId}"></div>` +
+      appendChildToBody(ad, createAPVTag() + insertVASTMethod(bidRequest.bidId, body.vastxml));
+  }
+  ad = appendChildToBody(ad, body.beacon);
+  if (removeWrapper(ad)) return removeWrapper(ad);
+  return ad;
+}
+
+function createNativeAd(body) {
+  let native = {};
+  if (body.native_ad && body.native_ad.assets.length > 0) {
+    const assets = body.native_ad.assets;
+    for (let i = 0, len = assets.length; i < len; i++) {
+      switch (assets[i].id) {
+        case 1:
+          native.title = assets[i].title.text;
+          break;
+        case 2:
+          native.image = {
+            url: assets[i].img.url,
+            height: assets[i].img.h,
+            width: assets[i].img.w,
+          };
+          break;
+        case 3:
+          native.icon = {
+            url: assets[i].img.url,
+            height: assets[i].img.h,
+            width: assets[i].img.w,
+          };
+          break;
+        case 4:
+          native.sponsoredBy = assets[i].data.value;
+          break;
+        case 5:
+          native.body = assets[i].data.value;
+          break;
+        case 6:
+          native.cta = assets[i].data.value;
+          break;
+      }
+    }
+    native.clickUrl = body.native_ad.link.url;
+    native.clickTrackers = body.native_ad.link.clicktrackers || [];
+    native.impressionTrackers = body.native_ad.imptrackers || [];
+  }
+  return native;
+}
+
+function appendChildToBody(ad, data) {
+  return ad.replace(/<\/\s?body>/, `${data}</body>`);
+}
+
+function createAPVTag() {
+  const APVURL = 'https://cdn.apvdr.com/js/VideoAd.min.js';
+  let apvScript = document.createElement('script');
+  apvScript.type = 'text/javascript';
+  apvScript.id = 'apv';
+  apvScript.src = APVURL;
+  return apvScript.outerHTML;
+}
+
+function insertVASTMethod(targetId, vastXml) {
+  let apvVideoAdParam = {
+    s: targetId
+  };
+  let script = document.createElement(`script`);
+  script.type = 'text/javascript';
+  script.innerHTML = `(function(){ new APV.VideoAd(${JSON.stringify(apvVideoAdParam)}).load('${vastXml.replace(/\r?\n/g, '')}'); })();`;
+  return script.outerHTML;
```

Comment:
can you explain what this is doing/why it's needed?

Reply:
VAST data, in our sever responses , is converted by CSS Sprites etc.